Add a places layer

jake-low opened the pull request:
This PR adds a "Places" layer containing OSM features tagged `place=*`, which represent countries, states, provinces, cities, towns, villages, neighborhoods, etc. Features in this layer are all Point geometries: they represent the approximate center of the feature.

I added this layer partly because it's useful to me, but also because it's simple and therefore a good testbed for two new schema ideas:
- The `population` column is an integer. This is the first foray into parsing OSM's string-valued tags into other data types.
- The `name` column contains the value of that tag in OSM, which (usually) represents the name in the primary local language. But this layer also contains a `names` column which is a `Map<str, str>` which contains all of the OSM tags whose keys start with `name:`, e.g. `name:left`, `name:fr`, and even `name:etymology:wikidata`. Keys in the map are the tag key with the `name:` prefix dropped, and values are the OSM tag value. So for example `tags.names.es` contains the value of `name:es` (the Spanish language name), if available. I also added `alt_names` and `official_names` maps to complement the `alt_name` and `official_name` columns.

jake-low committed:
Add a places layer

1ec5 commented on src/places.py:
```diff
@@ -0,0 +1,101 @@
+    (cities, towns, neighborhoods, etc) from OSM - anything tagged place=*
+    that has a name.
+
+    Most place features are mapped as nodes, but some are mapped as areas
+    (typically neighborhoods, islands, etc). In these cases we include the
+    area's centroid in the output dataset.
```
`place=archipelago`/`island`/`islet` sound like good candidates for a different layer about landforms. Similarly, `place=ocean` would be a good candidate for a water layer, and `place=square` for a layer about pedestrian infrastructure or perhaps public spaces.

jake-low committed:
Exclute some place=* values from places layer

jake-low committed:
Rename places layer to settlements

jake-low commented:
I renamed the layer to `settlements`, and changed it to only include `place=*` values that represent human settlements (city, town, etc) or parts of settlements (borough, neighbourhood, etc).

ianthetechie commented on src/settlements.py:
```diff
@@ -0,0 +1,84 @@
+        ("official_names", pyarrow.map_(pyarrow.string(), pyarrow.string())),
+        ("wikidata", pyarrow.string()),
+        ("wikipedia", pyarrow.string()),
+        ("population", pyarrow.uint32()),
```
Nice! Yeah, I think this is a good example of using a non-stringly typed value makes the data more friendly to consumers. Realistically, nobody will ever care about a population that's not parseable as an integer.

Since this is OSM and we're all about bikeshedding edge cases... have you considered the [population of Asia](https://www.openstreetmap.org/node/36966065#map=19/51.208697/89.234375&layers=N)?

```suggestion
        ("population", pyarrow.uint64()),
```

ianthetechie commented on src/settlements.py:
```diff
@@ -0,0 +1,84 @@
+            return tags.get(column_name)
+
+
+def tags_with_prefix(prefix, tags):
+    """
+    Returns a dict of all tags with the given prefix string; keys in the
+    dict will have the prefix dropped.
+    """
+    prefix_len = len(prefix)
+    return {k[prefix_len:]: v for (k, v) in tags if k.startswith(prefix)}
```
I like this helper function! Since we'll probably need some variation of it elsewhere, maybe we should create put it in another module (helpers? sounds cliche but I'm not very creative at the moment :P) so others can call it as a top level function?

jake-low replied:
Good call, thanks! I wrote this thinking it'd be used in other layers too but forgot to move it to a helpers module.

ianthetechie reviewed (commented):
Just a few minor nits. I like the way you're factoring out the map-style tags, the approach to alternate names, and the new name. Place is such an overloaded term; this captures it better I think!

jake-low committed:
Handle really big settlements

Co-authored-by: Ian Wagner <ian.wagner@stadiamaps.com>

jake-low committed:
Move tags_with_prefix() to a helpers.py module
